### PR TITLE
feat: add Anthropic support

### LINKED ISSUE #283: [Feature Request] Add support to Anthropic LLM
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [X] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

The Claude series from [Anthropic](https://www.anthropic.com/) company is one of the most popular LLMs and they are serious competitors of GPTs. They support a 100,000 tokens context window while still free for personal API usage.

### Solution

Add Claude-2 and Claude-instant-1 as backend models

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ocss884**
## Description

Describe your changes in detail.

## Motivation and Context

close #283
- [x] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Implemented Tasks

- [x] add Anthropic LLM (Claude-2, Claue-instant-1) backend
- [x] add an example of using Claude

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/utils/token_counting.py
Comment (Obs01ete) [inline camel/utils/token_counting.py:79]: Thank you for spotting. Did you discover a bug of sending the system prompt twice? We never observed it. Could you create a bug ticket to reproduce it and fix? We'd better fix the erroneous behavior in the main code than create a detour here in your code.

**Review by Obs01ete — CHANGES_REQUESTED** [review]
Thank you @ocss884. This is a very well prepared PR. Let's sort out what is wrong with the two system messages in the main code. Also we will need tests for the code that you added including examples. Please wrap the code that would crash if not Anthropic key provided into skipif as here:
https://github.com/camel-ai/camel/blob/3ba175450085b03d69ac1c68156c41d3ae843204/test/models/test_open_source_model.py#L61

Please run code coverage as in CONTRIBUTING.md and for all the code that you have added make sure that it covered.

File: camel/utils/token_counting.py
Comment (ocss884) [reply camel/utils/token_counting.py:79]: I think it is an expected behavior since in init_chat() the doc_string says "..sending the system message again to the agents using chat messages". You can spot it [here](https://github.com/camel-ai/camel/blob/3ba175450085b03d69ac1c68156c41d3ae843204/camel/societies/role_playing.py#L351C4-L351C4). 
I should say it is confusing for me since in `role_play`, this `init_chat` function is used to get the first `input_assistant_msg` and sometimes it is just a modified version of `user_sys_msg` (for example in the [open source example](https://github.com/camel-ai/camel/blob/3ba175450085b03d69ac1c68156c41d3ae843204/examples/open_source_models/role_playing_with_open_source_model.py#L60C9-L60C9)). I'm happy to help if you can confirm this is not desired.

File: camel/configs.py
Comment (dandansamax) [inline camel/configs.py:170]: Could you modify the "Args" docstrings to fit the code?

File: camel/configs.py
Comment (dandansamax) [inline camel/configs.py:188]:
```suggestion
ANTHROPIC_API_PARAMS = {
```

**Review by dandansamax — CHANGES_REQUESTED** [review]
Great code! Leave some comments.

**Comment by ocss884** [comment]
> Thank you @ocss884. This is a very well prepared PR. Let's sort out what is wrong with the two system messages in the main code. Also we will need tests for the code that you added including examples. Please wrap the code that would crash if not Anthropic key provided into skipif as here:
> 
> https://github.com/camel-ai/camel/blob/3ba175450085b03d69ac1c68156c41d3ae843204/test/models/test_open_source_model.py#L61
> 
> Please run code coverage as in CONTRIBUTING.md and for all the code that you have added make sure that it covered.

**Comment by ocss884** [comment]
Anthropic have a more openai-like 'Message completion' in beta now to replace their current `text completion`.
In the new function, people can use the user-assistant form, for example:
```
[
  {"role": "user", "content": "Hello there."},
  {"role": "assistant", "content": "Hi, I'm Claude. How can I help you?"},
  {"role": "user", "content": "Can you explain LLMs in plain English?"},
]
```
I will update once they are done.

**Comment by Wendong-Fan** [comment]
Hey @ocss884 , I plan to do the review later this week, would you like to solve the conflicts and update this PR with latest master branch? Thanks a lot!

**Comment by ocss884** [comment]
> Hey @ocss884 , I plan to do the review later this week, would you like to solve the conflicts and update this PR with latest master branch? Thanks a lot!

Hi @Wendong-Fan I just updated this PR. Due to an openai recent release, I limit the package version #474 . Some test cases are failed cuz this PR was opened when I was not a member and that makes the OPENAI_API_KEY secret empty (you can check the log). `pytest .` gives an all pass on my local.

File: camel/utils/commons.py
Comment (Wendong-Fan) [inline camel/utils/commons.py:59]:
```suggestion
        print(self)
        if self.model_type.is_openai:
            if 'OPENAI_API_KEY' not in os.environ:
                raise ValueError('OpenAI API key not found.')
            return func(self, *args, **kwargs)
        elif self.model_type.is_anthropic:
            if 'ANTHROPIC_API_KEY' not in os.environ:
                raise ValueError('Anthropic API key not found.')          
            return func(self, *args, **kwargs)
        else:
            raise ValueError('Unsupported model type.')
```

File: pyproject.toml
Comment (Wendong-Fan) [inline pyproject.toml:32]: why fix the maximal openai version here?

File: pyproject.toml
Comment (Wendong-Fan) [inline pyproject.toml:38]: move it into [tool.poetry.extras] part instead keep it in [tool.poetry.dependencies] to reduce default installation?

File: camel/models/anthropic_model.py
Comment (Wendong-Fan) [inline camel/models/anthropic_model.py:67]: it would be better to add docstring even it's sample function, if these functions is not internally used

**Review by Wendong-Fan — COMMENTED** [review]
Thanks for the contribution, overall looks great, left some comments, after your update we can merge it

File: pyproject.toml
Comment (ocss884) [reply pyproject.toml:32]: Originally it worked as a temporary fix for issue #474 . And yes the upper bound of openai package version is no longer needed now.

File: pyproject.toml
Comment (ocss884) [reply pyproject.toml:38]: I don't agree since we have a whole file describes how to treat anthropic models and it need anthropic package itself. Move the requirement as extra will simply break it. It is as import as the `openai` package
https://github.com/camel-ai/camel/pull/288/files#diff-f90260a7e58e0c09b9d2e737a580635d71a7310f1a4adebdb3b41323a960698cR17

**Comment by ocss884** [comment]
Going to add Claude3 support before merge it.

**Comment by ocss884** [comment]
@Wendong-Fan @dandansamax Hi Wendong and Tianqi. Claude3 is ready now. But it seems I cannot approve the PR myself. @dandansamax could you help take a look? Thanks!

**Comment by Appointat** [comment]
Hi, @ocss884 , I found this PR was merged without approval and the pipelines were not passed. It is the wrong approach, and it must be reverted. 

Hi, @Wendong-Fan . You have to see this, please. ❗

image

File: camel/utils/token_counting.py
Comment (lightaime) [inline camel/utils/token_counting.py:286]: It seems `messages_to_prompt ` does not support anthropic models yet. How does it work?

https://github.com/camel-ai/camel/blob/916ab5eb5946ef3eb68a108b6dc0110375a76a9a/camel/utils/token_counting.py#L39

File: camel/utils/token_counting.py
Comment (Wendong-Fan) [inline camel/utils/token_counting.py]: why did this part deleted? @ocss884 

cc @lightaime

**Pull request merged**
